evalset/multiopt_test_funcs.py

```diff
@@ -23,3 +23,118 @@ def do_evaluate(self, x):
+
+  def do_evaluate(self, x):
+    x1, x2 = x
+    return -(
+        .5 * numpy.exp(-10 * (.8 * (x1 + .8) ** 2 + .7 * (x2 + .8) ** 2)) +
+        .5 * numpy.exp(-8 * (.3 * (x1 + .8) ** 2 + .6 * (x2 - .3) ** 2)) +
+        .5 * numpy.exp(-9 * (.8 * x1 ** 2 + .7 * x2 ** 2)) +
+        .5 * numpy.exp(-9 * (.8 * (x1 - .3) ** 2 + .7 * (x2 + .8) ** 2)) +          
+        .5 * numpy.exp(-10 * (.8 * (x1 - .8) ** 2 + .7 * (x2 - .8)** 2))
+    )
+
+class MidDMixtureOfGaussians01(MultioptTestFunction):
+
+  def __init__(self, dim=8):
+    assert dim == 8
+    super(MidDMixtureOfGaussians01, self).__init__(dim)
+    self.bounds = lzip([-1] * self.dim, [1] * self.dim)
+    self.fmin = -0.50212691955
+    self.fmax = -0.00001997307
+    self.local_minima_loc = [(-0.2, -0.5, 0, 0, 0, 0, 0, 0),
+                             (0.8, 0.3, 0, 0, 0, 0, 0, 0)]
+
+  def do_evaluate(self, x):
+    x1, x2, x3, x4, x5, x6, x7, x8 = x
+    return -(
+        .5 * numpy.exp(-10 * (.8 * (x1 + .2) ** 2 + .7 * (x2 + .5) ** 2)) +
+        .5 * numpy.exp(-8 * (.3 * (x1 - .8) ** 2 + .6 * (x2 - .3) ** 2))
+    )
+
+
+class MidDMixtureOfGaussians02(MultioptTestFunction):
+
+  def __init__(self, dim=8):
+    assert dim == 8
+    super(MidDMixtureOfGaussians02, self).__init__(dim)
+    self.bounds = lzip([-1] * self.dim, [1] * self.dim)
+    self.fmin = -0.50016818373
+    self.fmax = -0.00004539993
+    self.local_minima_loc = [(-0.5, -0.5, -0.5, -0.5, -0.5, -0.5, -0.5, -0.5),
+                             (0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5)]
+
+  def do_evaluate(self, x):
+    mu1 = 0.5 * numpy.ones(8)
+    mu2 = -0.5 * numpy.ones(8)
+    return -(
+        0.5 * numpy.exp(-sum((x - mu1)**2)) +
+        0.5 * numpy.exp(-sum((x - mu2)**2))
+    )
+
+class CosineExponential(MultioptTestFunction):
+
+  def __init__(self, dim=2):
+    assert dim == 2
+    super(CosineExponential, self).__init__(dim)
+    self.bounds = lzip([-1] * self.dim, [1] * self.dim)
+    self.fmin = -1
+    self.fmax = 0.88659774910
+    self.local_minima_loc = [(0.0, 0.0),
+                             (0.76627855858, 0.76627855858),
+                             (0.76627855858, -0.76627855858),
+                             (-0.76627855858, 0.76627855858),
+                             (-0.76627855858, -0.76627855858)]
+
+  def do_evaluate(self, x):
+    x1, x2 = x
+    return -numpy.exp(-0.2 * (x1**2 + x2**2)) * numpy.cos(4 * x1) * numpy.cos(4 * x2)
+
+class SimpleQuadratic(MultioptTestFunction):
+
+  def __init__(self, dim=2):
+    assert dim >= 2
+    super(SimpleQuadratic, self).__init__(dim)
+    self.bounds = lzip([-1] * self.dim, [1] * self.dim)
+    self.fmin = -1
+    self.fmax = 0
+    self.local_minima_loc = [(-1.0, -1.0) + tuple([0] * (dim - 2)),
+                             (-1.0, 1.0) + tuple([0] * (dim - 2)),
+                             (1.0, -1.0) + tuple([0] * (dim - 2)),
+                             (1.0, 1.0) + tuple([0] * (dim - 2))]
```

Comment:
I'm a bit confused.  Is this a quadratic function with several local minima?  Shouldn't it only have 1?  Or is this just a sort of misnaming, where these aren't local minima, they are just boundary minima?